File: tests/algorithms/transpiler/FS/maths/geometric_mean.fs
// Generated 2025-08-12 08:17 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec abs (x: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable x = x
    try
        __ret <- if x < 0.0 then (-x) else x
        raise Return
        __ret
    with
        | Return -> __ret
and pow_int (``base``: float) (exp: int) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable ``base`` = ``base``
    let mutable exp = exp
    try
        let mutable result: float = 1.0
        let mutable i: int = 0
        while i < exp do
            result <- result * ``base``
            i <- i + 1
        __ret <- result
        raise Return
        __ret
    with
        | Return -> __ret
and nth_root (x: float) (n: int) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable x = x
    let mutable n = n
    try
        if x = 0.0 then
            __ret <- 0.0
            raise Return
        let mutable guess: float = x
        let mutable i: int = 0
        while i < 10 do
            let mutable denom: float = pow_int (guess) (n - 1)
            guess <- (((float (n - 1)) * guess) + (x / denom)) / (float n)
            i <- i + 1
        __ret <- guess
        raise Return
        __ret
    with
        | Return -> __ret
and round_nearest (x: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable x = x
    try
        if x >= 0.0 then
            let n: int = int (x + 0.5)
            __ret <- float n
            raise Return
        let n: int = int (x - 0.5)
        __ret <- float n
        raise Return
        __ret
    with
        | Return -> __ret
and compute_geometric_mean (nums: float array) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable nums = nums
    try
        if (Seq.length (nums)) = 0 then
            failwith ("no numbers")
        let mutable product: float = 1.0
        let mutable i: int = 0
        while i < (Seq.length (nums)) do
            product <- product * (_idx nums (int i))
            i <- i + 1
        if (product < 0.0) && (((((Seq.length (nums)) % 2 + 2) % 2)) = 0) then
            failwith ("Cannot Compute Geometric Mean for these numbers.")
        let mutable mean: float = nth_root (abs (product)) (Seq.length (nums))
        if product < 0.0 then
            mean <- -mean
        let possible: float = round_nearest (mean)
        if (pow_int (possible) (Seq.length (nums))) = product then
            mean <- possible
        __ret <- mean
        raise Return
        __ret
    with
        | Return -> __ret
and test_compute_geometric_mean () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let eps: float = 0.0001
        let m1: float = compute_geometric_mean (unbox<float array> [|2.0; 8.0|])
        if (abs (m1 - 4.0)) > eps then
            failwith ("test1 failed")
        let m2: float = compute_geometric_mean (unbox<float array> [|5.0; 125.0|])
        if (abs (m2 - 25.0)) > eps then
            failwith ("test2 failed")
        let m3: float = compute_geometric_mean (unbox<float array> [|1.0; 0.0|])
        if (abs (m3 - 0.0)) > eps then
            failwith ("test3 failed")
        let m4: float = compute_geometric_mean (unbox<float array> [|1.0; 5.0; 25.0; 5.0|])
        if (abs (m4 - 5.0)) > eps then
            failwith ("test4 failed")
        let m5: float = compute_geometric_mean (unbox<float array> [|-5.0; 25.0; 1.0|])
        if (abs (m5 + 5.0)) > eps then
            failwith ("test5 failed")
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        test_compute_geometric_mean()
        printfn "%g" (compute_geometric_mean (unbox<float array> [|-3.0; -27.0|]))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()

File: tests/algorithms/x/PHP/maths/geometric_mean.php
<?php
ini_set('memory_limit', '-1');
function _panic($msg) {
    fwrite(STDERR, strval($msg));
    exit(1);
}
function mochi_abs($x) {
  if ($x < 0.0) {
  return -$x;
}
  return $x;
}
function pow_int($base, $exp) {
  $result = 1.0;
  $i = 0;
  while ($i < $exp) {
  $result = $result * $base;
  $i = $i + 1;
};
  return $result;
}
function nth_root($x, $n) {
  if ($x == 0.0) {
  return 0.0;
}
  $guess = $x;
  $i = 0;
  while ($i < 10) {
  $denom = pow_int($guess, $n - 1);
  $guess = (floatval(($n - 1)) * $guess + $x / $denom) / (floatval($n));
  $i = $i + 1;
};
  return $guess;
}
function round_nearest($x) {
  if ($x >= 0.0) {
  $n = intval(($x + 0.5));
  return floatval($n);
}
  $n = intval(($x - 0.5));
  return floatval($n);
}
function compute_geometric_mean($nums) {
  if (count($nums) == 0) {
  _panic('no numbers');
}
  $product = 1.0;
  $i = 0;
  while ($i < count($nums)) {
  $product = $product * $nums[$i];
  $i = $i + 1;
};
  if ($product < 0.0 && fmod(count($nums), 2) == 0) {
  _panic('Cannot Compute Geometric Mean for these numbers.');
}
  $mean = nth_root(mochi_abs($product), count($nums));
  if ($product < 0.0) {
  $mean = -$mean;
}
  $possible = round_nearest($mean);
  if (pow_int($possible, count($nums)) == $product) {
  $mean = $possible;
}
  return $mean;
}
function test_compute_geometric_mean() {
  $eps = 0.0001;
  $m1 = compute_geometric_mean([2.0, 8.0]);
  if (mochi_abs($m1 - 4.0) > $eps) {
  _panic('test1 failed');
}
  $m2 = compute_geometric_mean([5.0, 125.0]);
  if (mochi_abs($m2 - 25.0) > $eps) {
  _panic('test2 failed');
}
  $m3 = compute_geometric_mean([1.0, 0.0]);
  if (mochi_abs($m3 - 0.0) > $eps) {
  _panic('test3 failed');
}
  $m4 = compute_geometric_mean([1.0, 5.0, 25.0, 5.0]);
  if (mochi_abs($m4 - 5.0) > $eps) {
  _panic('test4 failed');
}
  $m5 = compute_geometric_mean([-5.0, 25.0, 1.0]);
  if (mochi_abs($m5 + 5.0) > $eps) {
  _panic('test5 failed');
}
}
function main() {
  test_compute_geometric_mean();
  echo rtrim(json_encode(compute_geometric_mean([-3.0, -27.0]), 1344)), PHP_EOL;
}
main();
